File: articles/water-energy-material-conditions-of-life/python/growth_energy_model.py
"""
Growth and energy-throughput workflow.

Run:
    python python/growth_energy_model.py
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from water_energy_core import fit_exponential_growth


ARTICLE_DIR = Path(__file__).resolve().parents[1]
GROWTH_PATH = ARTICLE_DIR / "data" / "growth_observations.csv"


def main() -> None:
    """Fit exponential growth by condition."""

    df = pd.read_csv(GROWTH_PATH)

    rows = []

    for condition, group in df.groupby("condition"):
        fit = fit_exponential_growth(
            group["time_h"].to_numpy(dtype=float),
            group["abundance"].to_numpy(dtype=float),
        )

        rows.append(
            {
                "condition": condition,
                "growth_rate_per_h": fit.growth_rate_per_h,
                "initial_abundance": fit.initial_abundance,
                "doubling_time_h": fit.doubling_time_h,
                "r_squared_log_space": fit.r_squared_log_space,
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/water-energy-material-conditions-of-life/r/growth_energy_model.R
# Growth and energy-throughput workflow in R.

growth_path <- file.path("data", "growth_observations.csv")

if (!file.exists(growth_path)) {
  growth_path <- file.path("..", "data", "growth_observations.csv")
}

growth <- read.csv(growth_path)

fit_condition <- function(df) {
  fit <- lm(log(abundance) ~ time_h, data = df)

  r_est <- coef(fit)[["time_h"]]
  n0_est <- exp(coef(fit)[["(Intercept)"]])
  td_est <- log(2) / r_est

  data.frame(
    condition = unique(df$condition),
    growth_rate_per_h = r_est,
    initial_abundance = n0_est,
    doubling_time_h = td_est,
    r_squared_log_space = summary(fit)$r.squared
  )
}

fit_df <- do.call(
  rbind,
  lapply(split(growth, growth$condition), fit_condition)
)

print(round(fit_df, 5))
